iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp

```diff
@@ -0,0 +1,353 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-flow-canonicalize-dispatch-ids"
+
+/// This file contains patterns allowing canonicalization taking advantage  of
+/// flow dispatch ops. When the ID range is know we may be able to calculate the
+/// min/max of the derived values and may allow folding affine.min or ForOp.
+/// This file duplicate some of the logic from
+/// `linalg::AffineMinSCFCanonicalizationPattern`. In the future we could unify
+/// this code back by upstreaminng an interface to MLIR to represent generic ID
+/// ops.
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+/// Traverse `e` and return an AffineExpr where all occurrences of `dim` have
```

Comment:
Agree, this can be moved to core however it will require some kind of interfaceOp to be able to interact with the IREE specific ops. I don't mind doing this right now but I'm not convince at this point I have something stable enough. 

The way to have this in core in my opinion is to have an InterfaceOp "IdRange" (or a better name) allowing to query the range of a given op.
Do you see another solution to have this in core? Do you think this should be done at this point or do you expect that further change in IREE flow may affect what we are doing there.

Reply:
If you take this `dim` here, there is nothing that says this comes from an ID op. It is only working under the constraint that `dim` is in `[min, max)`. I am not saying move the entire pattern to core. Just this method as a utility. If it is in core already, just move it to a convenient place and expose it as a utility method. There are a bunch of such methods in MLIR.

Reply:
I see, I didn't realize this was an option as it seems like exposing helper functions but in this case it does seem reasonable.

Reply:
Sent a patch to move this into a helper in core:
https://reviews.llvm.org/D94826